src/hooks/useFetch.js

```diff
@@ -4,103 +4,86 @@
  */
 
 import React from 'react';
-import { Client as BugsnagClient } from 'bugsnag-react-native';
 
-import { getAuthHeader, AUTH_ERROR_CODES } from 'sussol-utilities';
-
-import { SETTINGS_KEYS } from '../settings';
-import { UIDatabase } from '../database';
-
-const { SYNC_SITE_NAME, SYNC_SITE_PASSWORD_HASH } = SETTINGS_KEYS;
-
-const CONNECTION_TIMEOUT_PERIOD = 10000;
-
-const ERROR_CODES = {
-  ...AUTH_ERROR_CODES,
-  EMPTY_RESPONSE: 'No records found',
-};
-
-const bugsnagClient = new BugsnagClient();
-
-class BugsnagError extends Error {
-  constructor(message, data, ...args) {
-    super(message, ...args);
-    bugsnagClient.notify(this, report => {
-      report.errorMessage = message;
-      report.metadata = data;
-    });
-  }
-}
-
-const getAuthorizationHeader = () => {
-  const username = UIDatabase.getSetting(SYNC_SITE_NAME);
-  const password = UIDatabase.getSetting(SYNC_SITE_PASSWORD_HASH);
-  return getAuthHeader(username, password);
-};
-
-const isValidError = error => Object.values(ERROR_CODES).includes(error?.message);
+const doIf = (ifThis, doThis, ...withThese) => (ifThis() ? doThis(...withThese) : null);
 
 /**
  * Custom hook to fetch data inside mounted component.
  *
- * Fetches data on component mount. If the calling component is unmounted or the timeout is reached,
+ * If the calling component is unmounted or the timeout is reached,
  * the fetch is aborted.
  */
-export const useFetch = (url, options = {}, timeout = CONNECTION_TIMEOUT_PERIOD) => {
-  const [data, setData] = React.useState(null);
-  const [error, setError] = React.useState(null);
-  const isMounted = React.useRef(true);
+export const useFetch = url => {
+  const _timer = React.useRef(null);
+  const _controller = React.useRef(null);
+
+  const _isMounted = React.useRef(false);
+  const _isBlocked = React.useRef(false);
+
+  const [isLoading, _setIsLoading] = React.useState(false);
+  const [response, _setResponse] = React.useState(null);
+  const [error, _setError] = React.useState(null);
+
+  const isMounted = () => _isMounted.current;
+  const isUnblocked = () => !_isBlocked.current;
+
+  const setIsLoading = _isLoading => doIf(isMounted, _setIsLoading, _isLoading);
+  const setResponse = _response => doIf(isMounted, _setResponse, _response);
+  const setError = _error => doIf(isMounted, _setError, _error);
```

Comment:
now that we're wrapping the state setters, their references are unstable and may cause unnecessary/hard to find re-renders - Maybe `doIf` becomes a factory for producing wrapped memoized callbacks? 🤣 `doIfMountedFactory` 🤣  


But hmm.. maybe we're holding onto the `doIf` now just because it's funny?

`const setIsLoading = _isLoading => doIf(isMounted, _setIsLoading, _isLoading);`

```
const setIsLoading = _isLoading => {
	if (isMounted) _setIsLoading(_isLoading)
};
```

Reply:
I think the `useDoIf` and `useProtectedState` hooks should cover this (using `useCallback` as we talked about earlier). Have tested with unmounting etc. and is working (but always a risk of weird cases in the wild when we are jestless!).